File: [0069][Easy][Sqrt(x)]/Sqrt(x).py
# solution 1, e and log,
# leetcode time     cost : 40 ms
# leetcode memory   cost : 13 MB 
# Time  Complexity: O(1)
# Space Complexity: O(1)
from math import e, log
class Solution:
    def mySqrt(self, x):
        if x < 2:
            return x
        
        left = int(e**(0.5 * log(x)))
        right = left + 1
        return left if right * right > x else right

File: [0069][Easy][Sqrt(x)]/Sqrt(x).java
// solution 1, e and log,
// leetcode time     cost : 1 ms
// leetcode memory   cost : 38.7 MB 
// Time  Complexity: O(1)
// Space Complexity: O(1)
class Solution {
  public int mySqrt(int x) {
    if (x < 2) return x;

    int left = (int)Math.pow(Math.E, 0.5 * Math.log(x));
    int right = left + 1;
    return (long)right * right > x ? left : right;
  }
}